web/src/composebox_typeahead.ts

```diff
@@ -674,9 +674,22 @@ export function get_person_suggestions(
     let groups: UserGroup[];
     if (opts.filter_groups_for_mention) {
         groups = user_groups.get_user_groups_allowed_to_mention();
-    } else if (opts.filter_groups_for_guests && !settings_data.user_can_access_all_other_users()) {
-        groups = user_groups.get_realm_user_groups().filter((group) => {
+    } else if (opts.filter_groups_for_dm) {
+        const can_access_all_users = settings_data.user_can_access_all_other_users();
+        groups = user_groups.get_all_realm_user_groups().filter((group) => {
             const group_members = group.members;
+            if (group_members.size > 20) {
+                // We do not want user trying to DM more than 20 users
+                // together.
+                return false;
+            }
+
+            if (can_access_all_users) {
+                return true;
+            }
+
+            // If user cannot access all other users, we only show groups
+            // whose all members can be accessed by the user.
```

Comment:
"only show groups, all of whose members can be accessed by the user".

And then this check is OK, since we'll know we're only doing it on groups with at most 20 users.